### PR TITLE
feat: send LLM reasoning to Langfuse

### PR DESCRIPTION
**Pull request by sunner**
## Summary

- Capture reasoning returned by LiteLLM stream chunks through `reasoning_content`, unsigned `thinking_blocks`, and provider-specific fallback fields.
- Preserve every unsigned thinking block in order and include a stable `content` plus `reasoning_content` schema in Langfuse generation output whenever reasoning exists, including reasoning-only responses with empty content.
- Keep retries attempt-safe by discarding buffered reasoning from an abandoned pre-content stream attempt, and skip Anthropic signed cumulative thinking snapshots to prevent duplicate reasoning.
- Preserve the existing string output shape when reasoning is absent and keep reasoning out of the learner response stream.

## Why

Reasoning-capable models can return useful diagnostic output separately from their final answer. The shared LLM wrapper previously discarded that output before finalizing its Langfuse generation, so operators could not inspect it. Retry and signature edge cases also need to preserve only the successful, non-duplicated reasoning sequence.

## Validation

- `/Users/sunner/src/ai-shifu/.venv/bin/python -m pytest tests/test_llm.py tests/test_langfuse.py -q` — 70 passed, 1 skipped (the existing LiteLLM 1.95 contract test skips because the shared environment still has LiteLLM 1.80.11)
- `python3 scripts/check_architecture_boundaries.py` — passed with no new violations
- `python3 scripts/check_dev_tools.py` — core tooling passed
- `ruff format --check flaskr/api/llm/__init__.py tests/test_llm.py` and `git diff --check` — passed
- Ruff baseline comparison — unchanged: 98 existing findings in the LLM module and 8 in its test file on both `origin/main` and this branch

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved handling of streamed reasoning content across supported providers.
  * Prevented reasoning text from appearing in the visible response while preserving it in generated output records.
  * Fixed retry handling to discard reasoning from failed streaming attempts.
  * Improved output tracking for reasoning-only and empty-content streams.

### CODE REVIEW

File: src/api/flaskr/api/llm/__init__.py [hunk 1]
```diff
@@ -181,6 +182,54 @@ def _attach_usage_output_text(
...
+            return value.get(key)
+        return getattr(value, key, None)
+
+    candidates: list[Any] = [
+        _get(delta, "reasoning_content"),
+        _get(delta, "reasoning"),
+    ]
+    thinking_blocks = _get(delta, "thinking_blocks")
+    if isinstance(thinking_blocks, list):
+        candidates.extend(_get(block, "thinking") for block in thinking_blocks)
```
[the remark this answers is not in the document]
Comment (sunner) [reply src/api/flaskr/api/llm/__init__.py]: Fixed in 5e487f3c4. All `thinking` entries from a delta's `thinking_blocks` are normalized and joined in order; the fallback test now covers multiple blocks.

File: src/api/flaskr/api/llm/__init__.py [hunk 2]
```diff
@@ -181,6 +182,54 @@ def _attach_usage_output_text(
...
+def _build_langfuse_llm_output(
+    response_text: str,
+    reasoning_text: str,
+) -> str | dict[str, str]:
+    if not reasoning_text:
+        return response_text
+    output: dict[str, str] = {}
+    if response_text:
+        output["content"] = response_text
+    output["reasoning_content"] = reasoning_text
```
[the remark this answers is not in the document]
Comment (sunner) [reply src/api/flaskr/api/llm/__init__.py]: Fixed in 5e487f3c4. Structured reasoning output now always includes `content`, including an empty string, and a reasoning-only regression test covers the stable schema.

File: src/api/flaskr/api/llm/__init__.py [hunk 3]
```diff
@@ -991,6 +1041,8 @@ def invoke_llm(
         for res in response:
             if start_completion_time is None:
                 start_completion_time = datetime.now()
+            if len(res.choices):
+                reasoning_text += _extract_reasoning_delta(res.choices[0].delta)
```
[the remark this answers is not in the document]
Comment (sunner) [reply src/api/flaskr/api/llm/__init__.py:1071]: Fixed in 25768e1f6. Pre-content reasoning chunks are now buffered per stream attempt and released only when that attempt produces content or completes successfully. If a retryable connection error abandons the attempt, its buffered reasoning is discarded, so Langfuse sees only the successful attempt. Added a regression test covering stale reasoning before a retry.

File: src/api/flaskr/api/llm/__init__.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (sunner) [reply src/api/flaskr/api/llm/__init__.py]: Fixed in 25768e1f6. Signed Anthropic thinking blocks are now treated as cumulative signature snapshots and skipped; unsigned thinking blocks remain supported. Added a regression case proving the signed snapshot does not duplicate previously streamed reasoning.
